components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java

```diff
@@ -323,7 +323,11 @@ private void revokeTokensOfBindingRef(AuthenticatedUser user, String tokenBindin
                                     user.getFederatedIdPName(), authenticatedUser.getFederatedIdPName())
                             && StringUtils.equalsIgnoreCase(user.getUserName(), authenticatedUser.getUserName())) {
                         revokeFederatedTokens(consumerKey, user, accessTokenDO, tokenBindingReference);
-                    } else if (StringUtils.equalsIgnoreCase(userId, authenticatedUser.getUserId())) {
+                    } else if (StringUtils.equalsIgnoreCase(userId, authenticatedUser.getAuthenticatedSubjectIdentifier().split("@")[0])) {
```

Comment:
Shall we add another condition to limit this logic for B2B scenario, as in the ideal federated user case userID should be there

Reply:
From the authenticated user, there is no such attribute to check whether a B2B scenario. In these else blocks, if the userId is same as the authenticated users userId, the token revocation should occur. But here userId is set to null. In order to complete the flow without any error, this workaround was suggested. Here if the authenticated subject identifer has a hint for the userId, then let the token to be revoked.